README.md
# LTN Tracker

Adds a GUI to display information about your [LTN](https://mods.factorio.com/mod/LogisticTrainNetwork) network.

### LTN Tracker displays:
1. Depot stops, their current status and assigned trains.
2. Stations with provided/requested items, current deliveries and their control signals.
3. The combined inventory of all stations, with detailed information for each item.
4. A history of recent deliveries.
5. Alerts for stations and trains.

The GUI can be switched on and off using a toggle button in the top left corner or a configurable hotkey.

Special thanks to Optera for creating LTN and for his continued support during the development of this mod.

Mod Portal: https://mods.factorio.com/mod/LogisticTrainNetwork
Forums: https://forums.factorio.com/viewtopic.php?f=190&t=64842


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.1.13
Date: 2019-06-03

  Bugfixes:
    - Fixed bug in on_configuration_changed that caused a crash when upgrading from 0.1.10 or 0.1.11.

---------------------------------------------------------------------------------------------------
Version: 0.1.12
Date: 2019-06-02

  Changes:
    - Improved layout of inventory tab details pane.

  Bugfixes:
    - Fixed various bugs in calculations for stats window.

---------------------------------------------------------------------------------------------------
Version: 0.1.11
Date: 2019-05-30

  Features:
    - Added additional information about open requests to inventory tab details.

  Bugfixes:
    - Fixed that network ID filter did not always work properly.
    - Fixed that statistics tab only displayed last hour for all time settings.

---------------------------------------------------------------------------------------------------
Version: 0.1.10
Date: 2019-05-23

  Changes:
    - Added state and network ID to station information panel.

  Bugfixes:
    - Fixed crash when opening station tab (#29).
---------------------------------------------------------------------------------------------------
Version: 0.1.9
Date: 2019-05-22

  Bugfixes:
    - Fixed that migration files could cause issues in multiplayer.
---------------------------------------------------------------------------------------------------
Version: 0.1.8
Date: 2019-05-21

  Bugfixes:
    - Removed debug code that would cause a crash.
---------------------------------------------------------------------------------------------------
Version: 0.1.7
Date: 2019-05-21

  Changes:
    - Added network ID column to station tab, instead of displaying it as part of control signals (#27).
    - Changed available time frames for graphs on statisitcs tab to 1h/5h/25h from 1h/10h/50h.

  Bugfixes:
    - Fixed trains being send to wrong depot when using train info panel (#28).
    - Fixed that station signals did not update properly.
    - Fixed that "Disable alerts for underloaded trains" setting was inverted when enabling LTNT for the first time.
    - Fixed that no time duration was selected when opening the stats tab for the first time.
---------------------------------------------------------------------------------------------------
Version: 0.1.6
Date: 2019-05-20

  Features:
    - Statistics tab:
      * shows graphs of available amount of items over time
      * up to six items can be plotted at the same time
      * time frames of one hour, ten hours and 50 hours are available

  Bugfixes:
    - Fixed crash that could occur seemingly at random.
    - Fixed that on large bases depots and deliveries were not updated properly.
---------------------------------------------------------------------------------------------------
Version: 0.1.5
Date: 2019-05-11

  Features:
    - Added sidebar window with additional information about selected LTN stops.
    - Added sidebar window with additional information about selected trains.

  Changes:
    - LTNT main window can no longer be positioned on the left side to make space for an additional sidebar window.
    - Filter icons on inventory tab panes now work for any number of filter categories.
---------------------------------------------------------------------------------------------------
Version: 0.1.4
Date: 2019-05-08

  Features:
    - Added a bar graph showing amount of provided items over the last hour (highly experimental).

  Changes:
    - Marked LTN Tracker as conflicting mod to prevent usage of both stable and beta version at the same time.

  Bugfixes:
    - too many to list
---------------------------------------------------------------------------------------------------
Version: 0.1.3
Date: 2019-05-07

  Features:
    - Added hotkey (default: R) to refresh window.
    - Added hotkey (default: CTRL+F) to focus filter box on station tab.
    - All displayed numbers follow the factorio style of three digits + suffix.

  Bugfixes:
    - Fixed a lot of issues with on_configuration_changed.
---------------------------------------------------------------------------------------------------
Version: 0.1.2
Date: 2019-05-06

  Features:
    - provided items on inventory tab stay visible after they run out

  Bugfixes:
    - ...
---------------------------------------------------------------------------------------------------
Version: 0.1.1
Date: 2019-05-06

  Features:
    - items on inventory tab are sorted
    - items on inventory tab can be filtered by category

  Bugfixes:
    - some, now down to a bit less than lots of bugs
---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 2019-05-05

  Features:
    - probably lots of bugs
    - maybe some desyncs

ltnt/const.lua
-- constant definitions
local CONST = {}

CONST.settings = {
  ui = {
    ["ltnt-show-button"] = true,
    ["ltnt-window-height"] = true,
    ["ltnt-refresh-interval"] = true,
  },
  proc = {["ltnt-history-limit"] = true},
  debug = {
    ["ltnt-debug-print"] = true,
    ["ltnt-debug-level"] = true,
  },
}

CONST.global = {
  mod_name = "LTN_Tracker", -- preliminary name
  mod_prefix = "ltnt",
  gui_events = {defines.events.on_gui_click, defines.events.on_gui_checked_state_changed, defines.events.on_gui_text_changed}, -- events handled by on_gui_event
  mod_name_ltn = "LogisticTrainNetwork",
  mod_name_ltnc = "LTN_Combinator",
  minimal_version_ltn = "01.10.03",
  current_version_ltn = "01.10.04",
}

CONST.styles = {
  font_color_black = {40, 39, 40}
}

-- ui_ctrl.lua
CONST.ui_ctrl = {
  refresh_delay = 60, -- shortest time in ticks between consecutive ui refreshes
}

-- data_processing.lua
CONST.proc = {
  fluid_tolerance = 0.1,
  stops_per_tick = 20,
  deliveries_per_tick = 20,
  trains_per_tick = 30,
  items_per_tick = 50,
}

-- UI layout
CONST.main_frame = {
	n_tabs = 5,
  button_width = 185,
  button_sprite_bare = "ltnt_bt_sprite",
  button_sprite_alert = "ltnt_bt_alert_sprite",
  button_highlight_style = "ltnt_tab_button_highlight",
  button_default_style =  "ltnt_tab_button"
}

CONST.depot_tab = {
	tab_index = 1,
  pane_width_left = 355,
	col_width_left = {325, 55, 50, 52, 180},

  pane_width_right = 580,
	col_width_right = {160, 190, 101},
  -- parked / error / on delivery
  color_dict = {{r=0,g=1,b=0}, {r=1,g=0,b=0}, {r=1,g=1,b=1}},
}

CONST.station_tab = {
	tab_index = 2,
  header_col_width = {207, 29, 38*5+27, 38*4+21, 38*6, 1},
  station_col_width = 195,
  item_table_col_count = {5, 4, 6},
  item_table_max_rows = {4, 4, 2},
}

CONST.inventory_tab = {
	tab_index = 3,
  item_table_column_count = 13,
  details_item_tb_col_count = 9,
  details_width = 400,
  details_tb_col_width_stations = {300, 45},
  details_tb_col_width_deliveries = {160, 25, 160}
}

CONST.history_tab = {
	tab_index = 4,
	header_col_width = {175, 175, 175, 35, 55, 152},
	col_width = {175, 175, 175, 40, 45, 145},
	n_columns = 6,
  n_cols_shipment = 5,
}

CONST.alert_tab = {
	tab_index = 5,
  frame_width = {410, 525},
  n_columns = {3, 4},
	col_width_l = {170, 20, 160},
	col_width_r = {130, 120, 140, 0},
}

-- misc stuff

CONST.train_error_state_dict = {
  ["residuals"] = {"error.train-residual-cargo"},
  ["timeout"] = {"error.train-timeout"},
}

-- LTN definitions, copied from LTN's control.lua
local ltn = {
  ISDEPOT = "ltn-depot",
  NETWORKID = "ltn-network-id",
  MINTRAINLENGTH = "ltn-min-train-length",
  MAXTRAINLENGTH = "ltn-max-train-length",
  MAXTRAINS = "ltn-max-trains",
  REQUESTED_THRESHOLD = "ltn-requester-threshold",
  REQUESTED_STACK_THRESHOLD = "ltn-requester-stack-threshold",
  REQUESTED_PRIORITY = "ltn-requester-priority",
  NOWARN = "ltn-disable-warnings",
  PROVIDED_THRESHOLD = "ltn-provider-threshold",
  PROVIDED_STACK_THRESHOLD = "ltn-provider-stack-threshold",
  PROVIDED_PRIORITY = "ltn-provider-priority",
  LOCKEDSLOTS = "ltn-locked-slots",
}
ltn.is_control_signal = {
  [ltn.ISDEPOT] = true,
  [ltn.NETWORKID] = true,
  [ltn.MINTRAINLENGTH] = true,
  [ltn.MAXTRAINLENGTH] = true,
  [ltn.MAXTRAINS] = true,
  [ltn.REQUESTED_THRESHOLD] = true,
  [ltn.REQUESTED_STACK_THRESHOLD] = true,
  [ltn.REQUESTED_PRIORITY] = true,
  [ltn.NOWARN] = true,
  [ltn.PROVIDED_THRESHOLD] = true,
  [ltn.PROVIDED_STACK_THRESHOLD] = true,
  [ltn.PROVIDED_PRIORITY] = true,
  [ltn.LOCKEDSLOTS] = true,
}
ltn.ctrl_signal_var_name_bool = {
  [ltn.ISDEPOT] = "isDepot",
  [ltn.NOWARN] = "noWarnings",
}
ltn.ctrl_signal_var_name_num = {
  [ltn.NETWORKID] = "network_id",
  [ltn.MINTRAINLENGTH] = "minTraincars",
  [ltn.MAXTRAINLENGTH] = "maxTraincars",
  [ltn.MAXTRAINS] = "trainLimit",
  [ltn.REQUESTED_THRESHOLD] = "requestThreshold",
  [ltn.REQUESTED_STACK_THRESHOLD] = "requestStackThreshold",
  [ltn.REQUESTED_PRIORITY] = "requestPriority",
  [ltn.PROVIDED_THRESHOLD] = "provideThreshold",
  [ltn.PROVIDED_STACK_THRESHOLD] = "provideStackThreshold",
  [ltn.PROVIDED_PRIORITY] = "providePriority",
  [ltn.LOCKEDSLOTS] = "lockedSlots",
}
ltn.error_color_lookup = {
  [-1]= "signal-white",
  [0] = "signal-white", -- this is a modification, used when entity becomes invalid
  [1] = "signal-red",
  [2] = "signal-pink",
}
ltn.error_string_lookup = {
  [-1] = {"error.stop-no-init"},
  [0] = {"error.stop-invalid"},
  [1] = {"error.stop-disabled"},
  [2] = {"error.stop-duplicate"},
}
CONST.ltn = ltn
return CONST